geo/src/algorithm/line_interpolate_point.rs

```diff
@@ -24,67 +27,76 @@ use crate::{
-        let r = [*fraction * v[0] + s[0], *fraction * v[1] + s[1]];
-        Coordinate { x: r[0], y: r[1] }.into()
+    fn line_interpolate_point(&self, fraction: T) -> Self::Output {
+        if (fraction > T::zero()) && (fraction < T::one()) {
+            // fraction between 0 and 1, return a point between start and end
+            let diff = self.end - self.start;
+            let r = self.start + diff * (fraction);
+            if r.x.is_finite() && r.y.is_finite() {
+                return Some(r.into());
+            } else {
+                return None;
+            }
+        } else if fraction <= T::zero() {
+            // negative fractions just return the start point
+            return Some(self.start_point());
+        } else if fraction >= T::one() {
+            // fraction above one is just the end point
+            return Some(self.end_point());
+        } else {
+            // fraction is nan
+            return None;
+        }
     }
 }
 
 impl<T> LineInterpolatePoint<T> for LineString<T>
 where
-    T: CoordinateType + Float + Zero + AddAssign + One,
+    T: CoordinateType + Float + AddAssign,
     Line<T>: EuclideanLength<T>,
     LineString<T>: EuclideanLength<T>,
 {
-    type Output = Point<T>;
-
-    fn line_interpolate_point(&self, fraction: &T) -> Self::Output {
-        let total_length = self.euclidean_length();
-        let fractional_length = total_length.clone() * *fraction;
-        let mut cum_length = T::zero();
-        let mut queue = Vec::new();
-        for line in self.lines() {
-            let length = line.euclidean_length();
-            queue.push((
-                cum_length.clone(),
-                cum_length.clone() + length.clone(),
-                length.clone(),
-                line.clone(),
-            ));
-            cum_length += length;
-        }
-        match queue.iter().find(|x| x.1 >= fractional_length) {
-            Some(x) => {
-                let line_frac = (fractional_length - x.0) / x.2;
-                (x.3).line_interpolate_point(&line_frac)
+    type Output = Option<Point<T>>;
+
+    fn line_interpolate_point(&self, fraction: T) -> Self::Output {
+        if (fraction > T::zero()) && (fraction < T::one()) {
+            // if fraction is outside these bounds the result is trivial
+            let total_length = self.euclidean_length();
+            let fractional_length = total_length.clone() * fraction;
+            let mut cum_length = T::zero();
+            for segment in self.lines() {
+                let length = segment.euclidean_length();
+                if cum_length + length >= fractional_length {
+                    let segment_fraction = (fractional_length - cum_length) / length;
+                    return segment.line_interpolate_point(segment_fraction);
+                }
+                cum_length += length;
             }
-            None => self.points_iter().last().unwrap(),
+            return None;
```

Comment:
hm... what would it mean if we got here? That `total_length.is_nan()`? Can we `debug_assert` that?

Any other cases?

Reply:
this one made me think - added a debug_assert for whether `fractional_length` is nan, or if the line has no coordinates `self.num_coords() == 0`